AdblockPlusSafari/AppDelegate.m

```diff
@@ -36,34 +49,154 @@ - (BOOL)application:(UIApplication *)application willFinishLaunchingWithOptions:
 
 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
 {
-  self.adblockPlus = [[AdblockPlus alloc] init];
-
-  if (!self.adblockPlus.activated) {
-    [self.adblockPlus reloadContentBlockerWithCompletion:nil];
-  }
+  self.adblockPlus = [[AdblockPlusExtras alloc] init];
+  self.firstUpdateTriggered = NO;
 
   if ([self.window.rootViewController isKindOfClass:[RootController class]]) {
     ((RootController *)self.window.rootViewController).adblockPlus = self.adblockPlus;
   }
-  
+
+  [application setMinimumBackgroundFetchInterval:BackgroundFetchInterval];
   return YES;
 }
 
-- (void)applicationWillResignActive:(UIApplication *)application {
+- (void)applicationWillResignActive:(UIApplication *)application
+{
 }
 
-- (void)applicationDidEnterBackground:(UIApplication *)application {
+- (void)applicationDidEnterBackground:(UIApplication *)application
+{
+  if (!self.adblockPlus.reloading) {
+    return;
+  }
+
+  __weak __typeof(self) wSelf = self;
+  __weak __typeof(application) wApplication = application;
+
+  self.backgroundTaskIdentifier = [application beginBackgroundTaskWithExpirationHandler:^{
+    [wSelf setBackgroundTaskIdentifier:UIBackgroundTaskInvalid withApplication:wApplication];
+  }];
 }
 
-- (void)applicationWillEnterForeground:(UIApplication *)application {
+- (void)applicationWillEnterForeground:(UIApplication *)application
+{
+  [self setBackgroundTaskIdentifier:UIBackgroundTaskInvalid withApplication:application];
 }
 
 - (void)applicationDidBecomeActive:(UIApplication *)application
 {
   [self.adblockPlus checkActivatedFlag];
+
+  if (!self.firstUpdateTriggered && !self.adblockPlus.updating && self.adblockPlus.lastUpdate == nil) {
```

Comment:
Our QA was complaining, that filter list is not updated right after the first start of the application. It make sense, since filter lists might be outdated for months.

self.adblockPlus.lastUpdate == nil ... It means that there any of possible previous update haven't been successful.